### PR TITLE
[NET-6462] Update gateway resources job for mesh gw v2

### PR DESCRIPTION
**Pull request by jm96441n**
Changes proposed in this PR:
- create gateway class and gateway class config for mesh gateway if mesh gateways are enabled

How I've tested this PR:
tests
:eye: 

How I expect reviewers to test this PR:
:eye: 

Checklist:
- [X] Tests added
- [ ] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: control-plane/subcommand/gateway-resources/command.go
Comment (jm96441n) [inline control-plane/subcommand/gateway-resources/command.go:405]: I _think_ the names/controller names here are correct, would be helpful for someone to confirm

File: control-plane/subcommand/gateway-resources/command.go
Comment (jm96441n) [inline control-plane/subcommand/gateway-resources/command.go:21]: we were double importing the `core/v1` dep as `corev1` and `v1` so I changed this to only import `corev1`

File: control-plane/subcommand/gateway-resources/command.go
Comment (jm96441n) [inline control-plane/subcommand/gateway-resources/command.go:445]: side effect of this change https://github.com/hashicorp/consul-k8s/pull/3273/files#r1406708181

File: control-plane/subcommand/gateway-resources/command.go
Comment (jm96441n) [inline control-plane/subcommand/gateway-resources/command.go:528]: we need to DeepCopy here to avoid copying the mutex that proto's have on them

File: control-plane/subcommand/gateway-resources/command.go
Comment (jm96441n) [inline control-plane/subcommand/gateway-resources/command.go:81]: make this configurable to it's easier to pass in a different filename while testing to test the unmarshal logic

File: control-plane/subcommand/gateway-resources/command_test.go
Comment (jm96441n) [inline control-plane/subcommand/gateway-resources/command_test.go:309]: Going to update this test and the example yaml above once the PR linked in the comment merges so we can add the resources as well

File: control-plane/subcommand/gateway-resources/command.go
Comment (NiniOak) [inline control-plane/subcommand/gateway-resources/command.go:179]: ~ Might be a good idea to differentiate config error message `c.UI.Error(fmt.Sprintf("Error loading api gw config: %s", err))`

File: control-plane/subcommand/gateway-resources/command.go
Comment (NiniOak) [inline control-plane/subcommand/gateway-resources/command.go:185]: ~ update error message `c.UI.Error(fmt.Sprintf("Error loading mesh gw config: %s", err))`

File: control-plane/subcommand/gateway-resources/command_test.go
Comment (NiniOak) [inline control-plane/subcommand/gateway-resources/command_test.go:341]: we should probably verify this warning is triggered https://github.com/hashicorp/consul-k8s/pull/3273/files#diff-5234182194c12306a3b7a58b33e6908998d552b2ba30d1f42cde158645864c55R367

**Review by NiniOak — COMMENTED** [review]
Looks good. Quick suggestion:
The current tests only validate 
- successfully loading a config and 
- not returning an error (doing nothing) when a config file is not found. 
We need one more test that returns an error for an invalid config file.

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [inline control-plane/subcommand/gateway-resources/command.go:157]: This only _happens_ to be for the mesh-gateway only today. In the future, the same config file will contain other gateway types as well, so I'd avoid including meshgw in the flag name.

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [reply control-plane/subcommand/gateway-resources/command.go:185]: Similarly here, the config file we're working with will contain/configure other gateway types in the future

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [reply control-plane/subcommand/gateway-resources/command.go:392]: yep yep

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [inline control-plane/subcommand/gateway-resources/command.go:411]:
```suggestion
					Kind:      "GatewayClass",
```

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [inline control-plane/subcommand/gateway-resources/command.go:408]: I'd probably make this as specific as possible to avoid any conflicts and make it self-documenting to the end user

```suggestion
				ControllerName: "consul-mesh-gateway-controller",
```

**Review by nathancoleman — CHANGES_REQUESTED** [review]
Nice work! Just a few forward-looking requests

File: control-plane/subcommand/gateway-resources/command.go
Comment (missylbytes) [inline control-plane/subcommand/gateway-resources/command.go:362]: Related to above comments, maybe this becomes loadConfig() and the legacy becomes loadConfigv1() or stays as  loadAPIGWResourceConfig()? (Not sold on those names).
Eventually everything will come as one config file.

Comment (jm96441n) [reply control-plane/subcommand/gateway-resources/command.go:362]: yeah, I'm down for either, it might make sense to keep the existing as `loadAPIGWResourceConfig` since that's really the only config it's loading but open to suggestions

File: control-plane/subcommand/gateway-resources/command_test.go
Comment (jm96441n) [reply control-plane/subcommand/gateway-resources/command_test.go:375]: done and done

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [reply control-plane/subcommand/gateway-resources/command.go]: make sense to me. this part is internal, so we can change it in the future. wanted to make sure the user-facing flags were updated first and foremost

File: control-plane/subcommand/gateway-resources/command_test.go
Comment (nathancoleman) [inline control-plane/subcommand/gateway-resources/command_test.go:394]: Might as well close the file here, even though it's for a test. There's also a useful alternative func for strings

```suggestion
	tmpdir := t.TempDir()
	file, err := os.CreateTemp(tmpdir, "config.yaml")
	if err != nil {
		t.Fatal(err)
	}
	defer file.Close()

	_, err = file.WriteString(fileContent)
	if err != nil {
		t.Fatal(err)
	}
```

**Review by nathancoleman — APPROVED** [review]
A few nitpicks but nothing to block merge.
Nice work! 🎉

File: control-plane/subcommand/gateway-resources/command.go
Comment (missylbytes) [reply control-plane/subcommand/gateway-resources/command.go]: We can leave for now, but that comment still needs to be updated.

File: control-plane/subcommand/gateway-resources/command_test.go
Comment (jm96441n) [reply control-plane/subcommand/gateway-resources/command_test.go:394]: good catch!

File: control-plane/subcommand/gateway-resources/command_test.go
Comment (jm96441n) [reply control-plane/subcommand/gateway-resources/command_test.go:385]: TIL, that's nifty

File: control-plane/subcommand/gateway-resources/command.go
Comment (jm96441n) [reply control-plane/subcommand/gateway-resources/command.go]: ugh, stupid case sensitive find and replaces

File: control-plane/subcommand/gateway-resources/command_test.go
Comment (NiniOak) [inline control-plane/subcommand/gateway-resources/command_test.go:299]: typo: Missing the h in mesh `TestRun_loadMesGatewayConfigs` -> `TestRun_loadMeshGatewayConfigs`

Comment (jm96441n) [reply control-plane/subcommand/gateway-resources/command_test.go:299]: good catch, these actually shouldn't even read `Mesh` anymore

File: control-plane/subcommand/gateway-resources/command.go
Comment (jm96441n) [inline control-plane/subcommand/gateway-resources/command.go:391]: going to modify this a bit to be `createV2GatewayClassAndClassConfigs` since we'll be creating more than just mesh gateway config in the future

**Review by missylbytes — APPROVED** [review]

**Review by NiniOak — APPROVED** [review]
🚀

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [inline control-plane/subcommand/gateway-resources/command.go:390]: To please the linter that's currently failing

```suggestion
// create the GatewayClassConfig and GatewayClass for the gateway.
```

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [inline control-plane/subcommand/gateway-resources/command.go:362]: To please the linter that's currently failing

```suggestion
// loadGatewayConfigs reads and loads the configs from `/consul/config/config.yaml`, if this file does not exist nothing is done.
```

File: charts/consul/templates/gateway-resources-clusterrole.yaml
Comment (jm96441n) [inline charts/consul/templates/gateway-resources-clusterrole.yaml:24]: we were missing some rbac rules here

File: charts/consul/templates/gateway-resources-configmap.yaml
Comment (jm96441n) [inline charts/consul/templates/gateway-resources-configmap.yaml:29]: this needs to be converted to json, the yaml parser fails because this value without the `toJson` ends up being a string representation of a go map, the `toYaml` function ends up throwing some errors because of how it represents the map

**Review by missylbytes — APPROVED** [review]

File: control-plane/subcommand/gateway-resources/command.go
Comment (nathancoleman) [reply control-plane/subcommand/gateway-resources/command.go:548]: Is this still required?

**Pull request merged**
